drv/cosmo-seq-server/src/main.rs

```diff
@@ -0,0 +1,522 @@
+        let seq = fmc_periph::Sequencer::new(token);
+        ringbuf_entry!(Trace::Startup {
+            early_power_rdbks: (&seq.early_power_rdbks).into(),
+        });
+        ringbuf_entry!(Trace::SetState {
+            prev: None, // dummy value
+            next: PowerState::A2,
+            why: StateChangeReason::InitialPowerOn,
+            now,
+        });
+
+        ServerImpl {
+            state: PowerState::A2,
+            jefe: Jefe::from(JEFE.get_task_id()),
+            sys: Sys::from(SYS.get_task_id()),
+            hf: HostFlash::from(HF.get_task_id()),
+            seq,
+        }
+    }
+
+    fn get_state_impl(&self) -> PowerState {
+        self.state
+    }
+
+    /// Logs a set of state registers, returning the state machine states
+    fn log_state_registers(&self) -> StateMachineStates {
+        let seq_api_status = (&self.seq.seq_api_status).into();
+        let nic_api_status = (&self.seq.nic_api_status).into();
+        ringbuf_entry!(Trace::RegStateValues {
+            seq_api_status,
+            seq_raw_status: (&self.seq.seq_raw_status).into(),
+            nic_api_status,
+            nic_raw_status: (&self.seq.nic_raw_status).into(),
+        });
+        StateMachineStates {
+            seq: seq_api_status.a0_sm,
+            nic: nic_api_status.nic_sm,
+        }
+    }
+
+    /// Logs a set of power good registers
+    fn log_pg_registers(&self) {
+        ringbuf_entry!(Trace::RegPgValues {
+            rail_pgs: (&self.seq.rail_pgs).into(),
+            rail_pgs_max_hold: (&self.seq.rail_pgs_max_hold).into(),
+        });
+    }
+
+    fn set_state_impl(
+        &mut self,
+        state: PowerState,
+        why: StateChangeReason,
+    ) -> Result<(), CpuSeqError> {
+        let now = sys_get_timer().now;
+        ringbuf_entry!(Trace::SetState {
+            prev: Some(self.state),
+            next: state,
+            why,
+            now,
+        });
+
+        use fmc_periph::A0Sm;
+        match (self.get_state_impl(), state) {
+            (PowerState::A2, PowerState::A0) => {
+                self.seq.power_ctrl.modify(|m| m.set_a0_en(true));
+                let mut okay = false;
+                // Wait 2 seconds for power-up
+                for _ in 0..200 {
+                    let state = self.log_state_registers();
+                    match state.seq {
+                        Ok(A0Sm::Done) => {
+                            okay = true;
+                            break;
+                        }
+                        Ok(A0Sm::Faulted) | Err(_) => {
+                            break;
+                        }
+                        _ => (),
+                    }
+                    hl::sleep_for(10);
+                }
+
+                if !okay {
+                    // We'll return to A2, leaving jefe and our local state
+                    // unchanged (since they're set after this block).
+                    self.log_pg_registers();
```

Comment:
It would be nice to know what seq API state and seq raw state we were in at the point of timeout, in addition to the PGs. Combining these pieces of information can help us pinpoint which rail(s) didn't come up or let us figure out what we are waiting on.

Reply:
Sure, I added a call to `self.log_state_registers()` here as well (and below) in 128d6f990aa3